README.md
# Toasta

Simple, Smooth & Elegant toast libary for Flutter on all Platform. Inspired by iOS. 
<a href="https://github.com/megatunger/toasta" rel="" target="_blank">Github Repo</a>

| iOS                        | Android                                                  |
| -------------------------- | ------------------------------------------------------------ |
| ![](https://github.com/megatunger/toasta/blob/master/demo/ios.gif?raw=true) | ![](https://github.com/megatunger/toasta/blob/master/demo/android.gif?raw=true) |


<a href="https://megatunger.com/toasta" rel="" target="_blank">![Demo](https://github.com/megatunger/toasta/blob/master/demo/web.png?raw=true)</a>

<h1 align="center"><a href="https://megatunger.com/toasta" rel="" target="_blank">Web Demo</a></h2>

## How to use

Step 1: Add to pubspec.yaml

```
flutter pub add toasta
```

Step 2: Wrap your `MaterialApp` to `ToastaContainer`

```dart
class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ToastaContainer( // Wrap your app by ToastaContainer
      child: MaterialApp(
        ...
      ),
    );
  }
}
```

Step 3: Start using

```dart
final toast = Toast(title: "Hello World", subtitle: "This is Toasta");
Toasta(context).toast(toast);
```



| Property                   | Description                                                  | Default              |
| -------------------------- | ------------------------------------------------------------ | -------------------- |
| height  `double`           | Height of toast                                              | 56                   |
| width  `double`            | Width of toast                                               | 70% of screen width  |
| darkMode `bool`            | Dark mode                                                    | false                |
| fadeInSubtitle `bool`      | Fade in animation for subtitle                               | false                |
| title `String` `Widget`    | The upper part of toast, you can passing both String or Widget to customize | null                 |
| subtitle `String` `Widget` | The lower part of toast, you can passing both String or Widget to customize | null                 |
| leading `Widget`           | Customize leading part of toast                              | null                 |
| trailing `Widget`          | Customize trailing part of toast                             | null                 |
| onTap    `Function`        | Callback on tap Toast                                        | null                 |
| onAppear    `Function`     | Callback after appeared Toast                                | null                 |
| onExit     `Function`      | Callback after disappeared Toast                             | null                 |
| status    `Enum`           | You can passing ToastStatus `success` `failed` `warning` `info` | null                 |
| backgroundColor `Color`    | Customize background color                                   | null                 |
| duration      `Duration`   | Duration of toast                                            | Duration(seconds: 3) |
| borderRadius  `BorderRadius`   | Border Radius of toast                                            | BorderRadius.all(Radius.circular(25.0)) |

## Support

Feel free to suggest features, reporting bugs or customize attributes.


CHANGELOG.md
## 1.0.0
- First Release
## 1.0.1
- Update README.md
## 1.0.2
- Add borderRadius property
- Fix width & height property not being passed
### 1.0.3
- Replace solid background with blur feels like iOS
- Change drag zone threshold
- Downgrade provider dependency for more compatible
### 1.0.4
- Fix incorrect shadow position on Web

test/toasta_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:toasta/toasta.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/toasta.dart
library toasta;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:toasta/provider.dart';
import 'package:toasta/toast_element.dart';

import 'model.dart';

export 'enum.dart';
export 'model.dart';

class ToastaContainer extends StatefulWidget {
  const ToastaContainer({Key? key, required this.child}) : super(key: key);

  final Widget child;
  @override
  _ToastaContainerState createState() => _ToastaContainerState();
}

class _ToastaContainerState extends State<ToastaContainer> {
  @override
  Widget build(BuildContext context) {
    return ChangeNotifierProvider<ToastaProvider>(
      create: (context) => ToastaProvider(),
      child: Consumer<ToastaProvider>(
        builder: (context, _toastaProvider, child) {
          Widget toastElement = Container();
          if (_toastaProvider.currentToasting != null) {
            toastElement =
                ToastElement(element: _toastaProvider.currentToasting!);
          }
          return MaterialApp(
              debugShowCheckedModeBanner: false,
              home: Stack(children: [widget.child, toastElement]));
        },
      ),
    );
  }
}

class Toasta {
  final BuildContext context;

  Toasta(this.context);

  toast(Toast toast) {
    ToastaProvider toastaProvider =
        Provider.of<ToastaProvider>(context, listen: false);
    toastaProvider.toast(toast);
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:toasta/toasta.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ToastaContainer(
      child: MaterialApp(
        title: "Toasta Demo",
        debugShowCheckedModeBanner: false,
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: MyHomePage(),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  var darkMode = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: darkMode ? Colors.black : Colors.white,
      body: SingleChildScrollView(
        child: Center(
          child: Container(
            width: double.infinity,
            constraints: const BoxConstraints(maxWidth: 500),
            child: Column(
                mainAxisAlignment: MainAxisAlignment.center,
                crossAxisAlignment: CrossAxisAlignment.center,
                mainAxisSize: MainAxisSize.max,
                children: [
                  const SizedBox(height: 32),
                  Padding(
                    padding: const EdgeInsets.symmetric(horizontal: 32.0),
                    child: Container(
                      width: double.infinity,
                      height: 48,
                      child: ElevatedButton(
                        style: ElevatedButton.styleFrom(
                            elevation: 2,
                            primary: darkMode ? Colors.white : Colors.black),
                        child: Text('Title & Subtitle',
                            style: TextStyle(
                                color: darkMode ? Colors.black : Colors.white)),
                        onPressed: () {
                          toastTitleSubtitle(context);
                        },
                      ),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.symmetric(
                        vertical: 8, horizontal: 32.0),
                    child: Container(
                      width: double.infinity,
                      height: 48,
                      child: ElevatedButton(
                        style: ElevatedButton.styleFrom(
                            elevation: 2,
                            primary: darkMode ? Colors.white : Colors.black),
                        child: Text('Without Fade In Subtitle',
                            style: TextStyle(
                                color: darkMode ? Colors.black : Colors.white)),
                        onPressed: () {
                          toastNoFadeIn(context);
                        },
                      ),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.symmetric(
                        vertical: 8, horizontal: 32.0),
                    child: Container(
                      width: double.infinity,
                      height: 48,
                      child: ElevatedButton(
                        style: ElevatedButton.styleFrom(
                            elevation: 2,
                            primary: darkMode ? Colors.white : Colors.black),
                        child: Text('Custom Duration',
                            style: TextStyle(
                                color: darkMode ? Colors.black : Colors.white)),
                        onPressed: () {
                          toastCustomDuration(context);
                        },
                      ),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.symmetric(
                        vertical: 8, horizontal: 32.0),
                    child: Container(
                      width: double.infinity,
                      height: 48,
                      child: ElevatedButton(
                        style: ElevatedButton.styleFrom(
                            elevation: 2, primary: Colors.green),
                        child: Text('Success',
                            style: TextStyle(color: Colors.white)),
                        onPressed: () {
                          toastSuccess(context);
                        },
                      ),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.symmetric(
                        vertical: 8, horizontal: 32.0),
                    child: Container(
                      width: double.infinity,
                      height: 48,
                      child: ElevatedButton(
                        style: ElevatedButton.styleFrom(
                            elevation: 2, primary: Colors.red),
                        child: Text('Failed',
                            style: TextStyle(color: Colors.white)),
                        onPressed: () {
                          toastFailed(context);
                        },
                      ),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.symmetric(
                        vertical: 8, horizontal: 32.0),
                    child: Container(
                      width: double.infinity,
                      height: 48,
                      child: ElevatedButton(
                        style: ElevatedButton.styleFrom(
                            elevation: 2, primary: Colors.yellow),
                        child: Text('Warning',
                            style: TextStyle(color: Colors.black)),
                        onPressed: () {
                          toastWarning(context);
                        },
                      ),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.symmetric(
                        vertical: 8, horizontal: 32.0),
                    child: Container(
                      width: double.infinity,
                      height: 48,
                      child: ElevatedButton(
                        style: ElevatedButton.styleFrom(
                            elevation: 2, primary: Colors.blue),
                        child:
                            Text('Info', style: TextStyle(color: Colors.white)),
                        onPressed: () {
                          toastInfo(context);
                        },
                      ),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.all(32.0),
                    child: Container(
                      width: double.infinity,
                      height: 48,
                      child: OutlinedButton(
                        style: OutlinedButton.styleFrom(
                          side: BorderSide(
                              width: 1,
                              color: darkMode ? Colors.white : Colors.black),
                        ),
                        child: Text(darkMode ? 'Light Mode' : 'Dark Mode',
                            style: TextStyle(
                                color: darkMode ? Colors.white : Colors.black)),
                        onPressed: () {
                          setState(() {
                            darkMode = !darkMode;
                          });
                        },
                      ),
                    ),
                  ),
                  SelectableText('Built by @megatunger',
                      style: Theme.of(context).textTheme.caption)
                ]),
          ),
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }

  void toastTitleSubtitle(BuildContext context) {
    final toast = Toast(
        title: "Lorem ipsum dolor sit amet, consectetur adipiscing elit",
        subtitle:
            "sed do eiusmod tempor incididunt ut labore et dolore magna aliqua",
        darkMode: darkMode);
    Toasta(context).toast(toast);
  }

  void toastNoFadeIn(BuildContext context) {
    final toast = Toast(
        title: "Lorem ipsum dolor sit amet, consectetur adipiscing elit",
        subtitle:
            "sed do eiusmod tempor incididunt ut labore et dolore magna aliqua",
        darkMode: darkMode,
        fadeInSubtitle: false);
    Toasta(context).toast(toast);
  }

  void toastSuccess(BuildContext context) {
    final toast = Toast(
        title: "Lorem ipsum dolor sit amet, consectetur adipiscing elit",
        subtitle:
            "sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris",
        status: ToastStatus.success,
        darkMode: darkMode);
    Toasta(context).toast(toast);
  }

  void toastFailed(BuildContext context) {
    final toast = Toast(
        title: "Lorem ipsum dolor sit amet, consectetur adipiscing elit",
        subtitle:
            "sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris",
        status: ToastStatus.failed,
        darkMode: darkMode);
    Toasta(context).toast(toast);
  }

  void toastWarning(BuildContext context) {
    final toast = Toast(
        title: "Lorem ipsum dolor sit amet, consectetur adipiscing elit",
        subtitle:
            "sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris",
        status: ToastStatus.warning,
        darkMode: darkMode);
    Toasta(context).toast(toast);
  }

  void toastInfo(BuildContext context) {
    final toast = Toast(
        title: "Lorem ipsum dolor sit amet, consectetur adipiscing elit",
        subtitle:
            "sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris",
        status: ToastStatus.info,
        darkMode: darkMode);
    Toasta(context).toast(toast);
  }

  void toastCustomDuration(BuildContext context) {
    final toast = Toast(
        title: "Lorem ipsum dolor sit amet, consectetur adipiscing elit",
        duration: const Duration(seconds: 5),
        darkMode: darkMode);
    Toasta(context).toast(toast);
  }
}


lib/toast_element.dart
import 'dart:async';
import 'dart:ui';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:marquee/marquee.dart';
import 'package:provider/provider.dart';
import 'package:toasta/provider.dart';

import 'enum.dart';
import 'model.dart';

class ToastElement extends StatefulWidget {
  const ToastElement({Key? key, required this.element}) : super(key: key);

  final Toast element;
  @override
  _ToastElementState createState() => _ToastElementState();
}

class _ToastElementState extends State<ToastElement>
    with TickerProviderStateMixin {
  static const blur = 32.0;
  late final AnimationController _startController = AnimationController(
    vsync: this,
    duration: const Duration(milliseconds: 350),
  );
  late final AnimationController _fadeController = AnimationController(
    vsync: this,
    duration: const Duration(milliseconds: 700),
  );
  late final Animation<Offset> _startOffsetFloat =
      Tween(begin: const Offset(0.0, -0.20), end: Offset.zero).animate(
    CurvedAnimation(
      parent: _startController,
      curve: Curves.easeOutQuint,
    ),
  );
  late final AnimationController _scaleController = AnimationController(
    duration: const Duration(milliseconds: 700),
    vsync: this,
  );
  late final Animation<double> _animation = CurvedAnimation(
    parent: _scaleController,
    curve: Curves.easeOutQuint,
  );
  late Timer disappearTimer;
  @override
  void initState() {
    _startController.forward().then((_) {
      if (widget.element.onAppear != null) {
        widget.element.onAppear!();
      }
    });
    _scaleController.forward();
    if (widget.element.fadeInSubtitle == false) {
      _fadeController.duration = const Duration(milliseconds: 0);
      _fadeController.forward();
    } else {
      Future.delayed(const Duration(milliseconds: 200), () {
        _fadeController.forward();
      });
    }
    disappearTimer = Timer(
        widget.element.duration != null
            ? widget.element.duration!
            : const Duration(seconds: 3), () {
      disappear();
    });
    super.initState();
  }

  void disappear() {
    _startController.reverse().then((value) {
      if (widget.element.onExit != null) {
        widget.element.onExit!();
      }
      ToastaProvider toastaProvider =
          Provider.of<ToastaProvider>(context, listen: false);
      toastaProvider.pop();
    });
  }

  @override
  void dispose() {
    _startController.dispose();
    _scaleController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SlideTransition(
      position: _startOffsetFloat,
      child: ScaleTransition(
        scale: _animation,
        alignment: FractionalOffset.topCenter,
        child: Align(
          alignment: FractionalOffset.topCenter,
          child: SafeArea(
            child: Container(
              margin: kIsWeb ? const EdgeInsets.only(top: 16) : null,
              child: ClipRRect(
                borderRadius: widget.element.borderRadius != null
                    ? widget.element.borderRadius!
                    : const BorderRadius.all(
                        Radius.circular(25.0 + 7.0),
                      ),
                child: BackdropFilter(
                  filter: ImageFilter.blur(sigmaX: blur, sigmaY: blur),
                  child: Container(
                    color: Colors.transparent,
                    child: Container(
                      decoration: BoxDecoration(
                          borderRadius: widget.element.borderRadius != null
                              ? widget.element.borderRadius!
                              : const BorderRadius.all(
                                  Radius.circular(25.0),
                                ),
                          boxShadow: widget.element.darkMode == true
                              ? []
                              : [
                                  BoxShadow(
                                    color: Colors.black.withOpacity(0.13),
                                    spreadRadius: 3,
                                    blurRadius: 20,
                                    offset: const Offset(0, 9),
                                  )
                                ]),
                      child: GestureDetector(
                        onVerticalDragUpdate: (details) {
                          int sensitivity = 8;
                          if (details.delta.dy > sensitivity) {
                          } else if (details.delta.dy < -sensitivity) {
                            disappearTimer.cancel();
                            disappear();
                          }
                        },
                        child: ElevatedButton(
                          style: ElevatedButton.styleFrom(
                            elevation: 0,
                            primary: widget.element.darkMode == true
                                ? Colors.grey.withOpacity(0.5)
                                : Colors.white.withOpacity(0.9),
                            shape: RoundedRectangleBorder(
                              borderRadius: BorderRadius.circular(32.0),
                            ),
                          ),
                          onPressed: () {
                            if (widget.element.onTap != null) {
                              widget.element.onTap!();
                            }
                          },
                          child: Padding(
                            padding: const EdgeInsets.symmetric(
                                horizontal: 8.0, vertical: 6),
                            child: SizedBox(
                              width: widget.element.width != null
                                  ? widget.element.width!
                                  : MediaQuery.of(context).size.width > 640
                                      ? MediaQuery.of(context).size.width * 0.4
                                      : MediaQuery.of(context).size.width * 0.7,
                              height: widget.element.height != null
                                  ? widget.element.height!
                                  : 56,
                              child: Row(
                                children: [
                                  if (widget.element.leading != null)
                                    widget.element.leading!,
                                  ...toastStatus(),
                                  Expanded(
                                    child: Column(
                                      crossAxisAlignment:
                                          CrossAxisAlignment.center,
                                      mainAxisAlignment:
                                          MainAxisAlignment.center,
                                      children: [
                                        toastTitle(),
                                        ...toastSubtitle(),
                                      ],
                                    ),
                                  ),
                                  if (widget.element.trailing != null)
                                    widget.element.trailing!,
                                ],
                              ),
                            ),
                          ),
                        ),
                      ),
                    ),
                  ),
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }

  Color toastBackgroundColor() {
    if (widget.element.backgroundColor != null) {
      return widget.element.backgroundColor!;
    }
    return widget.element.darkMode == true ? Colors.grey : Colors.white;
  }

  List<Widget> toastStatus() {
    if (widget.element.status != null) {
      Widget icon = Container();
      switch (widget.element.status) {
        case ToastStatus.failed:
          icon = const CircleAvatar(
              backgroundColor: Colors.red,
              child:
                  Icon(Icons.highlight_remove, size: 20, color: Colors.white));
          break;
        case ToastStatus.warning:
          icon = const CircleAvatar(
              backgroundColor: Colors.yellow,
              child:
                  Icon(Icons.warning_rounded, size: 20, color: Colors.black));
          break;
        case ToastStatus.success:
          icon = const CircleAvatar(
              backgroundColor: Colors.green,
              child: Icon(Icons.check_rounded, size: 20, color: Colors.white));
          break;
        case ToastStatus.info:
          icon = const CircleAvatar(
              backgroundColor: Colors.blue,
              child: Icon(Icons.info_outline, size: 20, color: Colors.white));
          break;
        default:
          break;
      }

      return [
        SizedBox(
          width: 32,
          height: 32,
          child: icon,
        ),
        const SizedBox(width: 10),
      ];
    }
    return [];
  }

  Widget toastTitle() {
    if (widget.element.title == null) {
      return Container();
    }
    return widget.element.title.runtimeType == String
        ? Expanded(
            child: Marquee(
              blankSpace: 32,
              fadingEdgeStartFraction: 0.05,
              fadingEdgeEndFraction: 0.05,
              style: const TextStyle(
                  color: Colors.black, fontWeight: FontWeight.w500),
              text: widget.element.title ?? '',
            ),
          )
        : widget.element.title;
  }

  List<Widget> toastSubtitle() {
    if (widget.element.subtitle == null) {
      return [Container()];
    }
    return [
      FadeTransition(
          opacity: _fadeController.drive(CurveTween(curve: Curves.easeInOut)),
          child: widget.element.subtitle.runtimeType == String
              ? Text(
                  widget.element.subtitle ?? '',
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  style: const TextStyle(
                      color: Colors.black, fontWeight: FontWeight.w300),
                )
              : widget.element.subtitle),
      const SizedBox(height: 8)
    ];
  }
}
